src/client/Microsoft.Identity.Client/ManagedIdentity/ManagedIdentityClient.cs

```diff
@@ -0,0 +1,53 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using Microsoft.Identity.Client.Extensibility;
+using System.Threading.Tasks;
+using System.Threading;
+using System.Linq;
+using Microsoft.Identity.Client.Http;
+using Microsoft.Identity.Client.Internal;
+using Microsoft.IdentityModel.Abstractions;
+using Microsoft.Identity.Client.Core;
+using Microsoft.Identity.Client.Utils;
+
+namespace Microsoft.Identity.Client.ManagedIdentity
+{
+    /// <summary>
+    /// Class to initialize a managed identity and identify the service.
+    /// Original source of code: https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/identity/Azure.Identity/src/ManagedIdentityClient.cs
+    /// </summary>
+    internal class ManagedIdentityClient
+    {
+        internal const string MsiUnavailableError =
+            "Authentication with managed identity is unavailable. No managed identity endpoint found.";
+
+        private readonly ManagedIdentitySource _identitySource;
+
+        public ManagedIdentityClient(RequestContext requestContext)
+        {
+            _identitySource = SelectManagedIdentitySource(requestContext);
+        }
+
+        private async Task<ManagedIdentityResponse> AuthenticateCoreAsync(AppTokenProviderParameters parameters,
+            CancellationToken cancellationToken)
+        {
+            return await _identitySource.AuthenticateAsync(parameters, cancellationToken).ConfigureAwait(false);
+        }
+
+        internal async Task<AppTokenProviderResult> AppTokenProviderImplAsync(AppTokenProviderParameters parameters)
+        {
+            ManagedIdentityResponse response = await AuthenticateCoreAsync(parameters, parameters.CancellationToken).ConfigureAwait(false);
+
+            return new AppTokenProviderResult() { AccessToken = response.AccessToken, ExpiresInSeconds = DateTimeHelpers.GetDurationFromNowInSeconds(response.ExpiresOn) };
+        }
+
+        // This method tries to create managed identity source for different sources, if none is created then defaults to IMDS.
+        private static ManagedIdentitySource SelectManagedIdentitySource(RequestContext requestContext)
+        {
+            return AppServiceManagedIdentitySource.TryCreate(requestContext) ?? 
+                throw new MsalClientException(MsalError.ManagedIdentityRequestFailed, MsiUnavailableError);
```

Comment:
are we not adding the default to IMDS if App service is not found?

Reply:
Yes that is in the next PR. Will remove it then.